### ISSUE TITLE
Feature proposal: dynamic state IDs based on an input variable

### ISSUE THREAD

**Issue by aaronsteers** [meltano/meltano#6164]
This was called out as "nice to have" but not delivered in the initial iteration of auto-generated state IDs for `meltano run`. 

For scenarios where the 'environment' actually builds to multiple deploy targets, such as one per customer or one per CI/CD build number or one per CI/CD branch name, it would be great to have a configurable and dynamic `state ID suffix` that could take one or more inputs, and use those inputs to ensure a unique set of STATE ID refs when running via `meltano run`.

Today state IDs are generated like:

```
${environment_name}:${tap-name}-to-${target-name}
```

But we could add some per-environment setting such as `state_id_suffix` or `deploy_target_key` which would concatenate with the base state ID calc:

```
${environment_name}:${tap-name}-to-${target-name}:${suffix_str}
```

Taking the branch names example, and an environment name called `cicd`, this could look like the following:

```
cicd:tap-github-to-target-snowflake:main
cicd:tap-github-to-target-snowflake:develop
cicd:tap-github-to-target-snowflake:staging
cicd:tap-github-to-target-snowflake:feat/my-cool-feature
cicd:tap-github-to-target-snowflake:fix/my-super-important-fix
```

Where each of these variations are basically populated with some input like `CI_BRANCH_REF`.

**Comment by s7clarke10** [comment]
Noting that this feature is related to this discussion : https://github.com/meltano/meltano/discussions/6273 .

**Comment by aaronsteers** [comment]
@s7clarke10 - This topic came up today in office hours. One new idea that came up in that session was that an alternative solution here would actually be to have a dynamic environment name - or perhaps a dynamic `environment_id` optional key in the yaml config. Since environment name/ID is a seed into the state ID values, this solution could "just work".

A sample definition where you want Meltano to behave as if each customer has its own 'environment' (along with separate state tracking) without having to define the environment `n` number of times (and `n` could be up to 100s or 1000s).

```yaml
environments:
  prod-by-cust:
    id: prod-cust-${CUSTOMER_ID}   # new/proposed environment key could use a dynamic expression
    # ...
```

Also, this has a nice affect of playing nicely with Meltano's built-in echoing of the environment name/ID at the start of invocation, something like:

```
meltano run tap-something target-something
INFO: Meltano environment 'prod-cust-1234` is now active.
...
```

It could also play nicely with the CI/CD use case where you want an environment seeded by branch name.

```yaml
environments:
  cicd:
    id: cicd-${CI_BRANCH_NAME}
    # ...
```

```
meltano run tap-something target-something
INFO: Meltano environment 'cicd-feat/3142-my-awesome-branch` is now active.
...
```



cc @edgarrmondragon, @tayloramurphy

**Comment by edgarrmondragon** [comment]
@aaronsteers I'm not a fan of not having a predictable environment name (or ID) and relying on variables that are not known to Meltano (and perhaps to the user) until a command is executed. If the goal is to give more control over auto-generated state IDs, the arguably simpler path of offering a `state_prefix` or similar key per environment is preferable. Wdyt?

**Comment by aaronsteers** [comment]
@edgarrmondragon - I can see pros and cons to the approach, for sure. One challenge is that we _are_ overloading the environment construct to create logically distinct and separate "sub environments" - because the state ID suffix in these use cases often needs to change in concert with other environment-level details, such as the target S3 bucket being different.

When the _outputs_ as well as the inputs are changed (or 'keyed') according to the env var context, other tasks in the pipeline will not reliably be able to find and process resources unless we have a way of keeping track of what version of the environment we are talking about.

I'm less inclined to go all-in with sub-environments or environment matrixes defined inline in meltano, so this is why I thought having a dynamic environment name was a cleaner approach that still respects the complex interdependencies which are at play when we incorporate dynamic outputs for specific steps of the workflow.

**Comment by tayloramurphy** [comment]
I'm inclined to agree with @edgarrmondragon here. Part of the problem statement is:

> want Meltano to behave as if each customer has its own 'environment' (along with separate state tracking)

It's not clear to me what other benefit we're getting by having the dynamic environment id other than the state-id behavior change. The only problem we're really trying to solve here is giving the user more control over the state-id and I don't understand how bringing that to the environment level solves it better.

**Comment by s7clarke10** [comment]
Thank you all for the feedback and discussion on this topic.

We had an additional use case come up today, ingesting data from multiple schemas in the same source database in parallel (note some of the table names could be the same across source database schemas, so they need to end up in different target schemas). Ideally we don't want to have to create another inherited tap again. I really want to run the same tap / target combo with just slightly different environment variables being passed.

Example Tables:

Database1.HR.Employee
Database1.HR.Department
Database1.Scott.Employee
Database1.Scott.Department

So if I wish to execute two data pipelines from Airflow at once using the same tap-oracle config. Currently I can do something like this example below (Note: the environment is a little more complex, I get env variables via Chamber from AWS SSM but you will get the idea).

`export target_snowflake_schema=oracle_hr; meltano elt tap-oracle target-snowflake --state-id=tap_oracle_to_sf_hr`
`export target_snowflake_schema=oracle_scott; meltano elt tap-oracle target-snowflake --state-id=tap_oracle_to_sf_scott`

As a side note I have a single Meltano Environment defined. But I really use the Meltano Environment capability. I have three different AWS Accounts (Dev, Test, Prod) which have a full copy of all resources e.g. Meltano in Containers (ECS), SSM, Airflow etc. We did before Meltano Environments were release (which I like by the way), but kept it because with separate AWS Accounts there is a level of isolation when we need to upgrade the infrastructure like AWS managed services like Airflow (MWAA), Aurora (for the system db) etc.

So I guess I'm leaning ( with my lack of use of Meltano Environments ) towards environment an environment variable being set which could be a SUFFIX or PREFIX to the State.

**Comment by s7clarke10** [comment]
Today my team have been discussing about loading tables in parallel from source. In production some of our tables need to be full_table load as there are deletes at source so we need a full_table load to detect the deletes.

Given the 100's of tables we need to load (some very large) we need to load in parallel, a serial load will not complete in time.

I therefore will need to kick of multiple meltano jobs for the same tap / target combination from Airflow. Again the problem of state comes into the mix in that we want to to uniquely identify the state for each instance run in parallel to make this possible. Currently we need to use `meltano elt` to achieve this outcome.

It seems a number of other users appear to be implementing / wanting parallel loads for their production loads.

https://meltano.slack.com/archives/C01UTUSP34M/p1622133666048500

https://meltano.slack.com/archives/CFG3C3C66/p1615312305119200

**Comment by cwe5590** [comment]
I have a very similar use case to that originally described by @s7clarke10. Would love to see a way to manage state_id with the run command so that I can start taking advantage of mappers

**Comment by aaronsteers** [comment]
@cwe5590 - Have you seen the specific proposal in this related issue?:

- #6370 

Happy to report @ReubenFrankel is working on a PR for this feature - which would be hugely valueable I think for many 'dynamic state ID' use cases.

**Comment by cwe5590** [comment]
@aaronsteers, I had not seen the specific proposal for this issue. That proposal looks great to me. Will keep an eye on it and help out where I can.

**Comment by tayloramurphy** [comment]
@ReubenFrankel can you update your PR to close this issue as well once the PR is merged? 😄

**Comment by s7clarke10** [comment]
This is looking very exciting. Thank you all for you contributed to this feature. Really looking forward to trying it out. Thanks
